internal/convert/registryv1.go

```diff
@@ -1,34 +1,146 @@
+	"github.com/operator-framework/rukpak/internal/util"
 )
 
 type RegistryV1 struct {
-	CSV    v1alpha1.ClusterServiceVersion
-	CRDs   []apiextensionsv1.CustomResourceDefinition
-	Others []unstructured.Unstructured
+	PackageName string
+	CSV         v1alpha1.ClusterServiceVersion
+	CRDs        []apiextensionsv1.CustomResourceDefinition
+	Others      []unstructured.Unstructured
 }
 
 type Plain struct {
 	Objects []client.Object
 }
 
+func RegistryV1ToPlain(rv1 fs.FS) (fs.FS, error) {
+	reg := RegistryV1{}
+	fileData, err := fs.ReadFile(rv1, filepath.Join("metadata", "annotations.yaml"))
+	if err != nil {
+		return nil, err
+	}
+	annotationsFile := registry.AnnotationsFile{}
+	if err := yaml.Unmarshal(fileData, &annotationsFile); err != nil {
+		return nil, err
+	}
+	reg.PackageName = annotationsFile.Annotations.PackageName
+
+	var objects []*unstructured.Unstructured
+	const manifestsDir = "manifests"
+
+	entries, err := fs.ReadDir(rv1, manifestsDir)
+	if err != nil {
+		return nil, err
+	}
+	for _, e := range entries {
+		if e.IsDir() {
+			return nil, fmt.Errorf("subdirectories are not allowed within the %q directory of the bundle image filesystem: found %q", manifestsDir, filepath.Join(manifestsDir, e.Name()))
+		}
+		fileData, err := fs.ReadFile(rv1, filepath.Join(manifestsDir, e.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		dec := apimachyaml.NewYAMLOrJSONDecoder(bytes.NewReader(fileData), 1024)
+		for {
+			obj := unstructured.Unstructured{}
+			err := dec.Decode(&obj)
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			if err != nil {
+				return nil, fmt.Errorf("read %q: %w", e.Name(), err)
+			}
+			objects = append(objects, &obj)
+		}
+	}
+
+	for _, obj := range objects {
+		obj := obj
+		switch obj.GetObjectKind().GroupVersionKind().Kind {
+		case "ClusterServiceVersion":
+			csv := v1alpha1.ClusterServiceVersion{}
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &csv); err != nil {
+				return nil, err
+			}
+			reg.CSV = csv
+		case "CustomResourceDefinition":
+			crd := apiextensionsv1.CustomResourceDefinition{}
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &crd); err != nil {
+				return nil, err
+			}
+			reg.CRDs = append(reg.CRDs, crd)
+		default:
+			reg.Others = append(reg.Others, *obj)
+		}
+	}
+
+	plain, err := Simple(reg)
+	if err != nil {
+		return nil, err
+	}
+
+	var manifest bytes.Buffer
+	for _, obj := range plain.Objects {
+		yamlData, err := yaml.Marshal(obj)
+		if err != nil {
+			return nil, err
+		}
+		if _, err := fmt.Fprintf(&manifest, "---\n%s\n", string(yamlData)); err != nil {
+			return nil, err
+		}
+	}
+
+	now := time.Now()
+	plainFS := fstest.MapFS{
+		".": &fstest.MapFile{
+			Data:    nil,
+			Mode:    fs.ModeDir | 0755,
+			ModTime: now,
+		},
+		"manifests": &fstest.MapFile{
+			Data:    nil,
+			Mode:    fs.ModeDir | 0755,
+			ModTime: now,
+		},
+		"manifests/manifest.yaml": &fstest.MapFile{
+			Data:    manifest.Bytes(),
```

Comment:
nit: how come we store the buffer data as bytes versus a string? My thinking was we would want to persist a string representation of the converted manifests to disk (in the tar.gz) for clients to access later.

Reply:
This is just what the MapFS implementation wants. It's supports any files, including binary files, hence using byte slices.